collection/collection.go

```diff
@@ -28,22 +28,217 @@ var (
+			})
+			if err != nil {
+				log.Debugw("updating dataset across all collections", "InitID", evt.InitID, "err", err)
+			}
+		}
+	case event.ETAutomationWorkflowStopped:
+		if evt, ok := e.Payload.(event.WorkflowStoppedEvent); ok {
+			err := c.set.UpdateOneAcrossAllCollections(ctx, evt.InitID, func(vi *dsref.VersionInfo) {
+				vi.RunStatus = evt.Status
+			})
+			if err != nil {
+				log.Debugw("updating dataset across all collections", "InitID", evt.InitID, "err", err)
+			}
+		}
+	case event.ETAutomationWorkflowCreated:
+		if wf, ok := e.Payload.(workflow.Workflow); ok {
+			err := c.set.UpdateOneAcrossAllCollections(ctx, wf.InitID, func(vi *dsref.VersionInfo) {
+				vi.WorkflowID = wf.WorkflowID()
+			})
+
+			if err != nil {
+				log.Debugw("updating dataset across all collections", "InitID", wf.InitID, "err", err)
+			}
+		}
+	case event.ETAutomationWorkflowRemoved:
+		if wf, ok := e.Payload.(workflow.Workflow); ok {
+			err := c.set.UpdateOneAcrossAllCollections(ctx, wf.InitID, func(vi *dsref.VersionInfo) {
+				vi.WorkflowID = ""
+			})
+
+			if err != nil {
+				log.Debugw("updating dataset across all collections", "InitID", wf.InitID, "err", err)
+			}
+		}
+	case event.ETFSICreateLink:
+		if link, ok := e.Payload.(event.FSICreateLink); ok {
+			c.set.UpdateOneAcrossAllCollections(ctx, link.InitID, func(vi *dsref.VersionInfo) {
+				vi.FSIPath = link.FSIPath
+			})
+		}
+	case event.ETFSIRemoveLink:
+		if change, ok := e.Payload.(event.FSIRemoveLink); ok {
+			c.set.UpdateOneAcrossAllCollections(ctx, change.InitID, func(vi *dsref.VersionInfo) {
+				vi.FSIPath = ""
+			})
+		}
+	}
+	return nil
+}
+
+// Set maintains lists of datasets, with each list scoped to a user profile.
 type Set interface {
 	List(ctx context.Context, pid profile.ID, lp params.List) ([]dsref.VersionInfo, error)
+	Get(ctx context.Context, pid profile.ID, initID string) (dsref.VersionInfo, error)
 }
 
 // WritableSet is an extension interface on Set that adds methods for
 // adding and removing items
 type WritableSet interface {
 	Set
-	Put(ctx context.Context, profileID profile.ID, items ...dsref.VersionInfo) error
+	// Put adds a list of `dsref.VersionInfo`s to a user's collection
+	Put(ctx context.Context, pid profile.ID, item ...dsref.VersionInfo) error
+	// Delete removes datasets from a user's collection
 	Delete(ctx context.Context, profileID profile.ID, initIDs ...string) error
+	// UpdateUsernameAcrossAllCollections updates all instances of a username
+	// in all collections
+	UpdateUsernameAcrossAllCollections(ctx context.Context, changedUsernameProfileIDString, newUsername string) error
+	// AddOneAcrossAllCollections adds a new dataset's information to all
+	// collections
+	AddOneAcrossAllCollections(ctx context.Context, add dsref.VersionInfo) error
+	// UpdateOneAcrossAllCollections updates a dataset's information in all
+	// collections
+	UpdateOneAcrossAllCollections(ctx context.Context, initID string, mutate func(vi *dsref.VersionInfo)) error
```

Comment:
These `AcrossAllCollections` methods are part of what makes this interface expansive. Pros to keeping these in the interface itself: no need to lock & unlock for each iteration. The con is new implementations need to implement a greater number methods.